test/graph_gen_transfer_tests_2.cc

```diff
@@ -0,0 +1,223 @@
+}
+
+TEST_CASE("graph_generator doesn't generate data transfer commands for the same buffer and range more than once PORTED NOCOMMIT",
+    "[graph_generator][command-graph]") {
+	dist_cdag_test_context dctx(2);
+
+	const range<1> test_range = {128};
+	auto buf0 = dctx.create_buffer(test_range);
+
+	SECTION("when used in the same task") {
+		const auto tid_a = dctx.device_compute<class UKN(task_a)>(test_range).discard_write(buf0, acc::one_to_one{}).submit();
+		// Both of theses are consumer modes, meaning that both have a requirement on the buffer range produced in task_a
+		const auto tid_b = dctx.master_node_host_task().read(buf0, acc::all{}).write(buf0, acc::all{}).submit();
+		CHECK(dctx.query().find_all(command_type::push, node_id(1)).count() == 1);
+		CHECK(dctx.query().find_all(command_type::await_push, node_id(0)).count() == 1);
+	}
+
+	SECTION("when used in the same task by different chunks on the same worker node") {
+		// NOCOMMIT TODO: We can't set number of chunks atm
+	}
+
+	SECTION("when used in consecutive tasks") {
+		auto buf1 = dctx.create_buffer(test_range);
+		const auto tid_a = dctx.device_compute<class UKN(task_a)>(test_range).discard_write(buf0, acc::one_to_one{}).submit();
+		const auto tid_b = dctx.master_node_host_task().read(buf0, acc::all{}).discard_write(buf1, acc::all{}).submit();
+		CHECK(dctx.query().find_all(command_type::push, node_id(1)).count() == 1);
+		CHECK(dctx.query().find_all(command_type::await_push, node_id(0)).count() == 1);
+		const auto tid_c = dctx.master_node_host_task().read(buf0, acc::all{}).read(buf1, acc::all{}).submit();
+		// Assert that the number of pushes / await_pushes hasn't changed
+		CHECK(dctx.query().find_all(command_type::push, node_id(1)).count() == 1);
+		CHECK(dctx.query().find_all(command_type::await_push, node_id(0)).count() == 1);
+	}
+
+	SECTION("when used in parallel tasks") {
+		const auto tid_a = dctx.device_compute<class UKN(task_a)>(test_range).discard_write(buf0, acc::one_to_one{}).submit();
+		const auto tid_b = dctx.master_node_host_task().read(buf0, acc::all{}).submit();
+		CHECK(dctx.query().find_all(command_type::push, node_id(1)).count() == 1);
+		CHECK(dctx.query().find_all(command_type::await_push, node_id(0)).count() == 1);
+		const auto tid_c = dctx.master_node_host_task().read(buf0, acc::all{}).submit();
+		// Assert that the number of pushes / await_pushes hasn't changed
+		CHECK(dctx.query().find_all(command_type::push, node_id(1)).count() == 1);
+		CHECK(dctx.query().find_all(command_type::await_push, node_id(0)).count() == 1);
+	}
+}
+
+// NOTE: This behavior changed between master/worker and distributed scheduling; we no longer consolidate pushes.
+//       In part this is because of the way data is being tracked (on a per-command last writer basis),
+//       however importantly it also enables better communication/computation overlapping in some cases.
+//       This behavior may change again in the future!
```

Comment:
I assume the possible future change in behaviour is why we keep the test case around?

Reply:
Yes, and for documentation purposes, basically.